.github/workflows/gen_matrix.pl

```diff
@@ -0,0 +1,191 @@
+#!/usr/bin/env swipl
+
+%% This script is used to generate an output matrix (in JSON format)
+%% that the github actions can import and use to generate various test
+%% configurations.
+%%
+%% There are three main axes: os, compiler, and solver.  The os and
+%% compiler axis have a vector of versions, one or more of which
+%% should be designated as the main version.  The solver axis has a
+%% solver version vector for each defined solver, where again at least
+%% one of each should be designated as the main solver version.
+%%
+%% Adding versions or identifiers to any axis is done with a simple
+%% declarative statement in the first section below.
+%%
+%% The output matrix defines a more efficient selection of tests than
+%% a full cartesian product (because the latter would uselessly test
+%% many redundant configurations).
+%%
+%%  1. Test the main version of each solver on the code build with the
+%%     main version of the compiler, for each operating system
+%%     (solver_main_compiler_main).
+%%
+%%  2. Test the other versions of compiled code with the main version
+%%     of each solver, for each operating system
+%%     (solver_main_compiler_all).
+%%
+%%  3. Test the other versions of each solver on the main compiler
+%%     version of the code, for each operating system
+%%     (solver_all_compiler_main).  Note that the tests run against
+%%     each solver independently, but a test run will test all
+%%     available solvers (and skip any unavailable solvers), so the
+%%     most efficient matrix varies each solver version independently
+%%     (but in parallel) for each test run until all solver versions
+%%     are exhausted.
+%%
+%% This script can be run by hand to see the generated JSON output
+%% using SWI Prolog:
+%%
+%%    $ swipl .github/workflows/gen_matrix.pl
+
+:- initialization(main, main).
+:- use_module(library(http/json)).
+:- use_module(library(clpfd)).
+
+%% Declare every Operating System that should be tested (separately in
+%% sequence)
+os(ubuntu).
+
+%% Declare every Compiler that should be used to build the code
+%% (separately in sequence)
+compiler(ghc).
+
+%% Declare every Solver that should be used (in parallel, not in
+%% sequence: one version of every solver is chosen for each test.
+solver(z3).
+solver(yices).
+solver(stp).
+solver(cvc4).
+solver(boolector).
+solver(abc).
+
+%% Specify the main version of each variable (i.e. the most interesting value of that variable.
+main_version(ubuntu, "ubuntu-latest").
+main_version(ghc, "8.10.7").
+main_version(z3, "4_8_14").
+main_version(yices, "2_6").
```

Comment:
What criteria are used to determine which version of a tool should be the main one? I had assumed that the most recent versions would be chosen as the main ones, but `yices` and `ghc`'s versions _aren't_ the most recent, unlike for other tools. If there is a reason for this discrepancy, it would be worth noting in the comments.

Reply:
Updated the comments to describe this.  Please feel free to re-address this if you'd like more clarity.

Reply:
Thanks. I think the thing in particular I am most confused about is the choice to use an old version of Yices. Can you leave a comment saying why that Yices version is preferred? Moreover, the current CI setup uses Yices 2.6.4, which doesn't appear to be included here (AFAICT).